src/firmware/app/control/trajectory_planner.h

```diff
@@ -0,0 +1,239 @@
+    // The initial speed at the start of the trajectory [m/s]
+    float initial_speed;
+    // The final speed at the end of the trajectory [m/s]
+    float final_speed;
+} FirmwareRobotPathParameters_t;
+
+/**
+ * Returns a planned trajectory with the list of guarantees based on the assumptions below
+ *
+ *  Key assumptions & guarantees of this planner are:
+ *  - Trajectories are time-optimal assuming INFINITE JERK capability of the robot
+ *
+ *  - No speed on the trajectory is larger than the 'max_allowable_speed' parameter
+ *
+ *  - No acceleration value between points on the trajectory can be larger than the
+ * 'max_allowable_acceleration' input parameter
+ *
+ *  - Assuming the grip-limit of the robot IS THE SAME AS THE MAX ACCELERATION then at no
+ * point on the path can the sum of centripetal and acceleration force be greater than
+ * 'max_allowable_acceleration'
+ *
+ *  Trajectory generation is done by assuming constant acceleration capability.
+ *      - The generator will assume maximum acceleration for the robot between each
+ * segment on the path. If using max acceleration breaches either the speed limit of the
+ * robot, or the centripetal acceleration limit defined by the curvature and robot speed -
+ * the planner will assume the max acceleration that will remain bellow this limit.
+ *
+ *     - To ensure the robot is capable of deccelerating to reach any sudden changes in
+ * the path, the trajectory is checked for backwards continuity
+ *
+ * @pre path_parameters.num_segments > 2
+ * @pre path_parameters.max_allowable_acceleration > 0
+ * @pre path_parameters.max_allowable_speed > 0
+ * @pre path_parameters.init_speed >= 0
+ * @pre path_parameters.final_speed >= 0
+ * @pre The trajectory.trajectory_elements[] array is pre-allocated ot handle up to the
+ * TRAJECTORY_PLANNER_MAX_NUM_ELEMENTS limit
+ *
+ * @param path_parameters [in] The data structure including important path parameters as
+ * defined by the FirmwareRobotPathParameters struct
+ * @param trajectory [out] The planned trajectory that follows robot dynamics limitations
+ * with the appropriate guarantees and assumptions outlines above
+ *
+ * @return The outcome of the trajectory generation. Returns OK if trajectory is valid.
+ */
+enum TrajectoryPlannerGenerationStatus
+app_trajectory_planner_generate_constant_arc_length_segmentation(
+    FirmwareRobotPathParameters_t path_parameters, Trajectory_t* trajectory);
+
+/**
+ * Returns a constant interpolation period (time) trajectory based on an input trajectory
+ *
+ * This function is intended to take a variable time trajectory (which is easier to
+ * calculate) then interpolate a constant period trajectory. This is valuable for discrete
+ * time motion controllers that operate only at constant delta-time intervals
+ *
+ * @pre variable_time_trajectory is a valid (obeys the physical limitation of the robot)
+ * trajectory
+ * @pre constant_period_trajectory is pre=allocated up the the limit specified by
```

Comment:
Another idea: For the arrays that we do expect to be pre-allocated to `TRAJECTORY_PLANNER_MAX_NUM_ELEMENTS`, we could always specify the array size in the parameter.

Eg. `foo(Trajectory_t[TRAJECTORY_PLANNER_MAX_NUM_ELEMENTS] constant_period_trajectory);`

This should also give us an error if the array is not pre-allocated to that size. However doing this is apparently up for debate as to whether it's good style or not because in theory developers could change the values and forget to update one or the other. Since we're using a `#define` we might be more safe.

https://stackoverflow.com/questions/5187224/why-should-i-declare-a-c-array-parameters-size-in-a-function-header

@garethellis0

Reply:
Changed all array* data to array data[TRAJECTORY_PLANNER_MAX_NUM_SEGMENTS]